pandas/tests/reshape/merge/test_merge.py

```diff
@@ -1526,6 +1526,17 @@ def test_merge_on_ints_floats_warning(self):
             result = B.merge(A, left_on='Y', right_on='X')
             assert_frame_equal(result, expected[['Y', 'X']])
 
+    def test_merge_incompat_infer_object(self):
+        # GH21119
+        df1 = DataFrame({'key': Series([True, False], dtype=object)})
+        df2 = DataFrame({'key': [True, False]})
```

Comment:
you need to have a NaN in the object dtype as well, this should test inferred and bool on both sides (with and w/o NaN).